File: app/models/domain.py
from datetime import datetime
from enum import StrEnum
from typing import Annotated, Literal
from uuid import UUID

from pydantic import (
    AliasChoices,
    BaseModel,
    ConfigDict,
    Field,
    model_validator,
)


class EvidenceTypeEnum(StrEnum):
    SOURCE = "SOURCE"
    RULE = "RULE"
    DERIVED = "DERIVED"


class JobStatusEnum(StrEnum):
    QUEUED = "QUEUED"
    PROCESSING = "PROCESSING"
    COMPLETED = "COMPLETED"
    FAILED_PARSING = "FAILED_PARSING"
    FAILED = "FAILED"
    CANCELLED = "CANCELLED"


class ProductStatusEnum(StrEnum):
    RAW = "RAW"
    VALIDATED = "VALIDATED"
    REVIEW_REQUIRED = "REVIEW_REQUIRED"
    UNCLASSIFIED_HUMAN_REVIEW = "UNCLASSIFIED_HUMAN_REVIEW"
    MANUAL_ENTRY = "MANUAL_ENTRY"
    PUBLISHED = "PUBLISHED"
    ARCHIVED = "ARCHIVED"


class CriticActionEnum(StrEnum):
    ACCEPT = "ACCEPT"
    RE_EXTRACT = "RE_EXTRACT"
    RE_NORMALIZE = "RE_NORMALIZE"
    ESCALATE_HUMAN = "ESCALATE_HUMAN"


class ValidationSeverityEnum(StrEnum):
    CRITICAL = "CRITICAL"
    WARNING = "WARNING"


class NormalizationMethodEnum(StrEnum):
    PINT = "PINT"
    REGEX = "REGEX"
    DICTIONARY = "DICTIONARY"
    LLM_FALLBACK = "LLM_FALLBACK"
    DERIVED_RULE = "DERIVED_RULE"
    MANUAL_ENTRY = "MANUAL_ENTRY"


class BoundingBox(BaseModel):
    model_config = ConfigDict(extra="forbid")

    page_number: Annotated[int, Field(ge=1)]
    top_pct: Annotated[float, Field(ge=0.0, le=100.0)]
    left_pct: Annotated[float, Field(ge=0.0, le=100.0)]
    width_pct: Annotated[float, Field(ge=0.0, le=100.0)]
    height_pct: Annotated[float, Field(ge=0.0, le=100.0)]


class Evidence(BaseModel):
    model_config = ConfigDict(extra="forbid")

    evidence_id: UUID
    evidence_type: EvidenceTypeEnum
    source_text: str | None = None
    page_number: Annotated[int, Field(ge=1)] | None = None
    bounding_box: BoundingBox | None = None
    rule_id: str | None = None
    parent_attribute_ids: list[UUID] | None = None
    confidence_score: Annotated[float, Field(ge=0.0, le=1.0)]
    is_verified: bool = False

    @model_validator(mode="after")
    def validate_lineage(self) -> "Evidence":
        if self.evidence_type is EvidenceTypeEnum.SOURCE:
            if not self.source_text or self.page_number is None or self.bounding_box is None:
                raise ValueError(
                    "SOURCE evidence requires source_text, page_number, and bounding_box"
                )
        elif self.evidence_type is EvidenceTypeEnum.RULE:
            if not self.rule_id:
                raise ValueError("RULE evidence requires rule_id")
        elif self.evidence_type is EvidenceTypeEnum.DERIVED:
            if not self.rule_id or not self.parent_attribute_ids:
                raise ValueError(
                    "DERIVED evidence requires rule_id and parent_attribute_ids"
                )
        return self


class ProcessingJob(BaseModel):
    model_config = ConfigDict(extra="forbid")

    job_id: UUID
    status: JobStatusEnum
    file_path: str
    total_pages: int | None = Field(default=None, ge=0)
    skus_found: int | None = Field(default=None, ge=0)
    created_at: datetime
    completed_at: datetime | None = None
    error_code: str | None = None
    error_message: str | None = None


class ProductAttribute(BaseModel):
    model_config = ConfigDict(
        extra="forbid",
        populate_by_name=True,
    )

    attribute_id: UUID
    canonical_key: str = Field(min_length=1)
    raw_key: str = Field(min_length=1)
    raw_value: str
    normalized_value: str | None = None
    numeric_value: float | None = None
    unit: str | None = None
    attribute_confidence: Annotated[float, Field(ge=0.0, le=1.0)] = 1.0
    normalization_method: NormalizationMethodEnum
    is_derived: bool = False
    is_human_locked: bool = Field(
        default=False,
        validation_alias=AliasChoices("is_human_locked", "is_locked"),
        serialization_alias="is_human_locked",
    )
    locked_state_hash: str | None = Field(
        default=None,
        validation_alias=AliasChoices("locked_state_hash", "locked_hash"),
        serialization_alias="locked_state_hash",
    )
    requires_human_review: bool = False
    evidence: Evidence
    evidence_id: UUID | None = None

    @model_validator(mode="after")
    def validate_evidence_reference(self) -> "ProductAttribute":
        if self.evidence_id is None:
            self.evidence_id = self.evidence.evidence_id
        elif self.evidence_id != self.evidence.evidence_id:
            raise ValueError("evidence_id must match the nested evidence object")
        return self

    @property
    def is_locked(self) -> bool:
        return self.is_human_locked


class ValidationReport(BaseModel):
    model_config = ConfigDict(extra="forbid")

    report_id: UUID
    attribute_id: UUID
    passed: bool
    rule_name: str = Field(min_length=1)
    severity: ValidationSeverityEnum
    error_message: str
    created_at: datetime | None = None


class CriticDecision(BaseModel):
    model_config = ConfigDict(extra="forbid")

    decision_id: UUID
    action: CriticActionEnum
    target_attributes: list[UUID] = Field(default_factory=list)
    rationale: str = Field(min_length=1)
    retry_count: int = Field(default=0, ge=0)
    created_at: datetime | None = None


class Product(BaseModel):
    model_config = ConfigDict(extra="forbid")

    product_id: UUID
    sku: str = Field(min_length=1)
    category_id: str = Field(min_length=1)
    status: ProductStatusEnum
    attributes: list[ProductAttribute]
    composite_confidence: Annotated[float, Field(ge=0.0, le=1.0)] = 0.0
    category_confidence: Annotated[float, Field(ge=0.0, le=1.0)] | None = None

File: web/src/types/domain.ts
/**
 * Types for Unilog Catalog Engine Domain and API Contracts.
 */

export type JobStatus =
  | "PENDING"
  | "QUEUED"
  | "PROCESSING"
  | "REVIEW_REQUIRED"
  | "UNCLASSIFIED_HUMAN_REVIEW"
  | "COMPLETED"
  | "FAILED_PARSING"
  | "FAILED"
  | "CANCELLED";

export type ProductStatus =
  | "RAW"
  | "VALIDATED"
  | "REVIEW_REQUIRED"
  | "UNCLASSIFIED_HUMAN_REVIEW"
  | "MANUAL_ENTRY"
  | "PUBLISHED"
  | "ARCHIVED";

export type EvidenceType = "SOURCE" | "RULE" | "DERIVED";

export type NormalizationMethod =
  | "PINT"
  | "REGEX"
  | "DICTIONARY"
  | "LLM_FALLBACK"
  | "DERIVED_RULE"
  | "MANUAL_ENTRY";

export interface BoundingBox {
  page_number: number;
  top_pct: number;
  left_pct: number;
  width_pct: number;
  height_pct: number;
}

export interface Evidence {
  evidence_id: string;
  evidence_type: EvidenceType;
  source_text?: string | null;
  page_number?: number | null;
  bounding_box?: BoundingBox | null;
  rule_id?: string | null;
  parent_attribute_ids?: string[] | null;
  confidence_score: number;
  is_verified?: boolean;
}

export interface AttributeRecord {
  attribute_id: string;
  product_id?: string;
  canonical_key: string;
  raw_key: string;
  raw_value: string;
  normalized_value?: string | null;
  numeric_value?: number | null;
  unit?: string | null;
  attribute_confidence: number;
  normalization_method: NormalizationMethod | string;
  is_derived: boolean;
  is_human_locked: boolean;
  locked_state_hash?: string | null;
  requires_human_review: boolean;
  evidence_id?: string;
  evidence_type?: string;
  source_text?: string | null;
  page_number?: number | null;
  bounding_box_json?: string | null;
  rule_id?: string | null;
  parent_attribute_ids_json?: string | null;
  evidence_confidence?: number;
  evidence_is_verified?: boolean;
  evidence?: Evidence;
}

export interface ProductRecord {
  product_id: string;
  job_id: string;
  sku: string;
  category_id: string;
  status: ProductStatus;
  composite_confidence: number;
  category_confidence?: number | null;
  created_at: string;
  attributes?: AttributeRecord[];
}

export interface ProcessingJob {
  job_id: string;
  status: JobStatus;
  file_path: string;
  total_pages?: number | null;
  skus_found?: number | null;
  created_at: string;
  completed_at?: string | null;
  error_code?: string | null;
  error_message?: string | null;
  graph_state_json?: string | null;
}

export interface AuditLog {
  audit_id: string;
  job_id?: string | null;
  user_id: string;
  attribute_id?: string | null;
  action: string;
  previous_value?: string | null;
  new_value?: string | null;
  reason_code: string;
  timestamp: string;
  details_json?: string | null;
}

export interface ReviewQueueItem {
  job_id: string;
  status: JobStatus;
  file_path: string;
  created_at: string | null;
}

export interface ModifiedAttributePayload {
  attribute_id?: string | null;
  canonical_key?: string | null;
  raw_key?: string | null;
  raw_value?: string | null;
  normalized_value?: string | null;
  numeric_value?: number | null;
  unit?: string | null;
  is_human_locked?: boolean;
  locked_state_hash?: string | null;
  evidence_id?: string | null;
}

export type ReviewAction = "ACCEPT_AND_SUBMIT" | "SAVE_DRAFT";

export interface ReviewSubmitRequest {
  job_id: string;
  action: ReviewAction;
  modified_attributes?: ModifiedAttributePayload[];
  audit_reason?: string | null;
}

export interface ReviewSubmitResponse {
  status: string;
  job_id: string;
  audit_log_id?: string | null;
  revalidation_triggered: boolean;
}

export interface JobDetailResponse {
  job_id: string;
  status: JobStatus;
  file_path: string;
  total_pages?: number | null;
  skus_found?: number | null;
  created_at: string | null;
  completed_at?: string | null;
  state: Record<string, unknown>;
}

export interface JobUploadResponse {
  job_id: string;
  status: JobStatus;
  file_path: string;
  created_at: string;
}
